### PR TITLE
Let apply_function method return cube-like objects; custom update function

### PR DESCRIPTION
**Pull request by nmearl**
This PR adds the ability to return cube-like objects from the `apply_function` method, and also allows for providing a custom progress update function à la #454.

### CODE REVIEW

**Comment by keflavich** [comment]
These changes look fine, so I'll merge once we resolve test issues.

However, `apply_function` appears to be narrower in scope than its name applies.  `apply_function`, prior to this PR, could reduce the dimensionality of a cube entirely (to a scalar) or reduce it by 1, to a plane.  It could not return 3D or 1D results.  After this PR, it can return 3D results, but still not 1D.  Perhaps we should expand it in that direction?

File: spectral_cube/spectral_cube.py [hunk 1]
```diff
@@ -785,6 +786,13 @@ def apply_function(self, function, axis=None, weights=None, unit=None,
         progressbar : bool
             Show a progressbar while iterating over the slices/rays through the
             cube?
+        keep_shape : bool
+            If `True`, the returned object will be the same dimensionality as
+            the cube.
+        update_function : function
+            An alternative tracker for the progress of applying the function
+            to the cube data. If `progressbar` is `True`, this argument is
```
Comment (keflavich) [inline spectral_cube/spectral_cube.py]: ``progressbar``

File: spectral_cube/spectral_cube.py [hunk 2]
```diff
@@ -791,7 +791,7 @@ def apply_function(self, function, axis=None, weights=None, unit=None,
             the cube.
         update_function : function
             An alternative tracker for the progress of applying the function
-            to the cube data. If `progressbar` is `True`, this argument is
+            to the cube data. If progressbar is true, this argument is
```
Comment (keflavich) [inline spectral_cube/spectral_cube.py]: You don't like using double-backticks?  I realize my comment before just looked like ``progressbar``; I was trying to show \`\`progressbar\`\`

Comment (nmearl) [reply spectral_cube/spectral_cube.py]: Hah, got it! I was wondering if that's what you meant, but figured there was some sphinx flag that just wouldn't allow formatting in docstrings for whatever reason.

Comment (keflavich) [reply spectral_cube/spectral_cube.py]: Yeah, I think this should work now.  Let's see....

**Comment by keflavich** [comment]
Yeah, I think we've fixed the issue that was keeping CI from completing.  Let's see....

**Pull request merged**